tools/chpl-language-server/src/chpl-language-server.py

```diff
@@ -0,0 +1,331 @@
+        elif decl.intent() == "type":
+            return SymbolKind.TypeParameter
+        else:
+            return SymbolKind.Variable
+    return None
+
+
+def get_symbol_information(
+    decl: chapel.core.NamedDecl, uri: str
+) -> Optional[SymbolInformation]:
+    loc = Location(uri, location_to_range(decl.location()))
+    if kind := decl_kind(decl):
+        # TODO: should we use DocumentSymbol or SymbolInformation
+        # LSP spec says prefer DocumentSymbol, but nesting doesn't work out of the box. implies that we need some kind of visitor pattern to build a DS tree
+        # using symbol information for now, as it sort-of autogets the tree structure
+        is_deprecated = chapel.is_deprecated(decl)
+        return SymbolInformation(loc, decl.name(), kind, deprecated=is_deprecated)
+    return None
+
+
+@dataclass
+class NodeAndRange:
+    node: chapel.core.AstNode
+    rng: Range = field(init=False)
+
+    def __post_init__(self):
+        self.rng = location_to_range(self.node.location())
+
+    def get_location(self):
+        path = self.node.location().path()
+        return Location(f"file://{path}", self.rng)
+
+
+@dataclass
+class ResolvedPair:
+    ident: NodeAndRange
+    resolved_to: NodeAndRange
+
+
+@dataclass
+class FileInfo:
+    uri: str
+    context: chapel.core.Context
+    segments: List[ResolvedPair] = field(default_factory=list)
+    siblings: chapel.SiblingMap = field(init=False)
+
+    def __post_init__(self):
+        self.rebuild_segments()
+
+    def parse_file(self) -> List[chapel.core.AstNode]:
+        """
+        Given a file URI, return the ASTs making up that file. Advances
+        the context if one already exists to make sure an updated result
+        is returned.
+        """
+
+        return self.context.parse(self.uri[len("file://") :])
+
+    def get_asts(self) -> List[chapel.core.AstNode]:
+        with self.context.track_errors() as _:
+            return self.parse_file()
+
+    def rebuild_segments(self):
+        asts = self.get_asts()
+        # get ids
+        self.segments = [
+            ResolvedPair(NodeAndRange(node), NodeAndRange(to))
+            for node, _ in chapel.each_matching(asts, chapel.core.Identifier)
+            if (to := node.to_node())
+        ]
+        self.segments.sort(key=lambda s: s.ident.rng.start)
+        self.siblings = chapel.SiblingMap(self.get_asts())
+
+
+def run_lsp():
+    """
+    Start a language server on the standard input/output, and use it to
+    report linter warnings as LSP diagnostics.
+    """
+
+    server = LanguageServer("chplcheck", "v0.1")
+
+    contexts: Dict[str, FileInfo] = {}
+
+    def get_context(uri: str, do_update: bool = False) -> Tuple[FileInfo, List[Any]]:
+        """
+        The LSP driver maintains one Chapel context per-file. This is to avoid
+        having to reset all files' text etc. when a single file is updated.
+        There may be a more principled approach we can take in the future.
```

Comment:
File text I'd imagine would be re-loaded from disk, and subsequent queries will not be re-run thanks to memoization magic ™️ . The main reason only one context per file exists is because Dyno crashes and burns on conflicting file names today. At least that was my impression -- I don't recall thinking too hard about re-setting file text.

Reply:
I rewrote this docstring, let me know what you think.